lib/models/message_container.dart

```diff
@@ -0,0 +1,181 @@
+      for (var packMessage in messages) {
+        final update = result.update(
+          message: packMessage,
+          receivedTimestamp: receivedTimestamp,
+          source: source,
+        );
+        if (update != null) result = update;
+      }
+      return result;
+    }
+    // update pack only if new data differ from saved ones
+    return switch (message.runtimeType) {
+      LocationMessage => locationMessage != null &&
+              locationMessage!.containsEqualData(message as LocationMessage)
+          ? null
+          : copyWith(
+              locationMessage: message as LocationMessage,
+              lastMessageRssi: rssi,
+              lastUpdate:
+                  DateTime.fromMillisecondsSinceEpoch(receivedTimestamp),
+              source: source,
+            ),
+      BasicIDMessage => basicIdMessage != null &&
+              basicIdMessage!.containsEqualData(message as BasicIDMessage)
+          ? null
+          : copyWith(
+              basicIdMessage: message as BasicIDMessage,
+              lastMessageRssi: rssi,
+              lastUpdate:
+                  DateTime.fromMillisecondsSinceEpoch(receivedTimestamp),
+              source: source,
+            ),
+      SelfIDMessage => selfIdMessage != null &&
+              selfIdMessage!.containsEqualData(message as SelfIDMessage)
+          ? null
+          : copyWith(
+              selfIdMessage: message as SelfIDMessage,
+              lastMessageRssi: rssi,
+              lastUpdate:
+                  DateTime.fromMillisecondsSinceEpoch(receivedTimestamp),
+              source: source,
+            ),
+      OperatorIDMessage => operatorIdMessage != null &&
+              operatorIdMessage!.containsEqualData(message as OperatorIDMessage)
+          ? null
+          : copyWith(
+              operatorIdMessage: message as OperatorIDMessage,
+              lastMessageRssi: rssi,
+              lastUpdate:
+                  DateTime.fromMillisecondsSinceEpoch(receivedTimestamp),
+              source: source,
+            ),
+      AuthMessage => authenticationMessage != null &&
+              authenticationMessage!.containsEqualData(message as AuthMessage)
+          ? null
+          : copyWith(
+              authenticationMessage: message as AuthMessage,
+              lastMessageRssi: rssi,
+              lastUpdate:
+                  DateTime.fromMillisecondsSinceEpoch(receivedTimestamp),
+              source: source,
+            ),
+      SystemMessage => systemDataMessage != null &&
+              systemDataMessage!.containsEqualData(message as SystemMessage)
+          ? null
+          : copyWith(
+              systemDataMessage: message as SystemMessage,
+              lastMessageRssi: rssi,
+              lastUpdate:
+                  DateTime.fromMillisecondsSinceEpoch(receivedTimestamp),
+              source: source,
+            ),
+      _ => null
+    };
+  }
+
+  pigeon.MessageSource getPackSource() => source;
```

Comment:
This one might be a getter instead? The checks below might be as well if you really want.

Reply:
yes i will use getters